File: tests/runners/typescript/tests/test_09_sqlite.ts
// Test Case 09: SQLite Accessor
// Tests SQLite database accessor type generation
// Type-check only - validates generated types are correct

import { TestSqlite } from '../generated/09_sqlite/typescript/schema';
import { UserTable, PostTable, CommentTable, SqliteDb, DbTable } from '../generated/09_sqlite/typescript/schema_sqlite_accessor';

// Test entity type creation
function testEntityCreation(): void {
    console.log('  Testing entity creation...');

    const user: TestSqlite.User = {
        id: 1,
        name: 'TestUser',
        email: 'test@example.com',
        createdAt: 1700000000
    };

    console.assert(user.id === 1, 'user.id');
    console.assert(user.name === 'TestUser', 'user.name');
    console.assert(user.email === 'test@example.com', 'user.email');
    console.assert(user.createdAt === 1700000000, 'user.createdAt');

    const post: TestSqlite.Post = {
        id: 1,
        userId: 1,
        title: 'Test Title',
        content: 'Test Content'
    };

    console.assert(post.id === 1, 'post.id');
    console.assert(post.userId === 1, 'post.userId');
    console.assert(post.title === 'Test Title', 'post.title');

    const comment: TestSqlite.Comment = {
        id: 1,
        postId: 1,
        userId: 2,
        content: 'Test Comment'
    };

    console.assert(comment.id === 1, 'comment.id');
    console.assert(comment.postId === 1, 'comment.postId');
    console.assert(comment.userId === 2, 'comment.userId');

    console.log('    PASS');
}

// Test PostStatus enum
function testEnum(): void {
    console.log('  Testing PostStatus enum...');

    console.assert(TestSqlite.PostStatus.Draft === 0, 'PostStatus.Draft should be 0');
    console.assert(TestSqlite.PostStatus.Published === 1, 'PostStatus.Published should be 1');
    console.assert(TestSqlite.PostStatus.Archived === 2, 'PostStatus.Archived should be 2');

    console.log('    PASS');
}

// Test optional fields
function testOptionalFields(): void {
    console.log('  Testing optional fields...');

    // User with optional email undefined
    const userNoEmail: TestSqlite.User = {
        id: 2,
        name: 'Bob',
        createdAt: 1700000001
    };
    console.assert(userNoEmail.email === undefined, 'email should be undefined');

    // Post with optional content undefined
    const postNoContent: TestSqlite.Post = {
        id: 2,
        userId: 1,
        title: 'No Content Post'
    };
    console.assert(postNoContent.content === undefined, 'content should be undefined');

    console.log('    PASS');
}

// Test table accessor types
function testTableAccessorTypes(): void {
    console.log('  Testing table accessor types...');

    // These are type checks - we can't actually run SQLite in type-check mode
    // but we can verify the types compile correctly
    const userTable: UserTable = new UserTable();
    const postTable: PostTable = new PostTable();
    const commentTable: CommentTable = new CommentTable();

    // Verify DbTable<T> properties
    console.assert(typeof userTable.length === 'number', 'length should be number');
    console.assert(typeof userTable.isEmpty === 'boolean', 'isEmpty should be boolean');
    console.assert(Array.isArray(userTable.all), 'all should be array');

    console.log('    PASS');
}

// Test SqliteDb type
function testSqliteDbType(): void {
    console.log('  Testing SqliteDb type...');

    const db: SqliteDb = new SqliteDb();

    // Verify table accessors exist
    console.assert(db.users instanceof UserTable, 'users should be UserTable');
    console.assert(db.posts instanceof PostTable, 'posts should be PostTable');
    console.assert(db.comments instanceof CommentTable, 'comments should be CommentTable');

    // Verify isOpen property
    console.assert(typeof db.isOpen === 'boolean', 'isOpen should be boolean');
    console.assert(db.isOpen === false, 'isOpen should be false initially');

    console.log('    PASS');
}

// Main
console.log('=== Test Case 09: SQLite Accessor ===');
testEntityCreation();
testEnum();
testOptionalFields();
testTableAccessorTypes();
testSqliteDbType();
console.log('=== All tests passed! ===');

File: tests/runners/cpp/tests/test_09_sqlite.cpp
// Test Case 09: SQLite Schema Types
// Tests the generated schema types for SQLite tables
// Note: This test does not use actual SQLite - it only validates types

#include <iostream>
#include <cassert>
#include <optional>

#include "schema.hpp"
#include "schema_loaders.hpp"

using namespace test::sqlite;

void test_user_creation() {
    std::cout << "  Testing User creation..." << std::endl;

    User user;
    user.id = 1;
    user.name = "TestUser";
    user.email = "test@example.com";
    user.created_at = 1700000000;

    assert(user.id == 1);
    assert(user.name == "TestUser");
    assert(user.email.has_value());
    assert(user.email.value() == "test@example.com");
    assert(user.created_at == 1700000000);

    std::cout << "    PASS" << std::endl;
}

void test_post_creation() {
    std::cout << "  Testing Post creation..." << std::endl;

    Post post;
    post.id = 1;
    post.user_id = 1;
    post.title = "Test Title";
    post.content = "Test Content";

    assert(post.id == 1);
    assert(post.user_id == 1);
    assert(post.title == "Test Title");
    assert(post.content.has_value());
    assert(post.content.value() == "Test Content");

    std::cout << "    PASS" << std::endl;
}

void test_comment_creation() {
    std::cout << "  Testing Comment creation..." << std::endl;

    Comment comment;
    comment.id = 1;
    comment.post_id = 1;
    comment.user_id = 2;
    comment.content = "Test Comment";

    assert(comment.id == 1);
    assert(comment.post_id == 1);
    assert(comment.user_id == 2);
    assert(comment.content == "Test Comment");

    std::cout << "    PASS" << std::endl;
}

void test_optional_fields() {
    std::cout << "  Testing optional fields..." << std::endl;

    // User with no email
    User user;
    user.id = 2;
    user.name = "Bob";
    user.created_at = 1700000001;

    assert(!user.email.has_value());

    // Post with no content
    Post post;
    post.id = 2;
    post.user_id = 1;
    post.title = "No Content Post";

    assert(!post.content.has_value());

    std::cout << "    PASS" << std::endl;
}

void test_post_status_enum() {
    std::cout << "  Testing PostStatus enum..." << std::endl;

    assert(static_cast<int32_t>(PostStatus::Draft) == 0);
    assert(static_cast<int32_t>(PostStatus::Published) == 1);
    assert(static_cast<int32_t>(PostStatus::Archived) == 2);

    PostStatus status = PostStatus::Published;
    assert(status == PostStatus::Published);

    std::cout << "    PASS" << std::endl;
}

void test_binary_serialization() {
    std::cout << "  Testing binary serialization..." << std::endl;

    User original;
    original.id = 12345;
    original.name = "Binary Test User";
    original.email = "binary@test.com";
    original.created_at = 1700000123;

    // Serialize
    std::vector<uint8_t> buffer;
    {
        polygen::BinaryWriter writer(buffer);
        polygen_loaders::write_User(writer, original);
    }

    // Deserialize
    User loaded;
    {
        polygen::BinaryReader reader(buffer);
        loaded = polygen_loaders::read_User(reader);
    }

    assert(loaded.id == original.id);
    assert(loaded.name == original.name);
    assert(loaded.email.has_value());
    assert(loaded.email.value() == original.email.value());
    assert(loaded.created_at == original.created_at);

    std::cout << "    PASS (serialized " << buffer.size() << " bytes)" << std::endl;
}

void test_binary_optional_null() {
    std::cout << "  Testing binary serialization with null optional..." << std::endl;

    User original;
    original.id = 99;
    original.name = "No Email";
    // email is std::nullopt by default
    original.created_at = 1700099999;

    // Serialize
    std::vector<uint8_t> buffer;
    {
        polygen::BinaryWriter writer(buffer);
        polygen_loaders::write_User(writer, original);
    }

    // Deserialize
    User loaded;
    {
        polygen::BinaryReader reader(buffer);
        loaded = polygen_loaders::read_User(reader);
    }

    assert(loaded.id == original.id);
    assert(loaded.name == original.name);
    assert(!loaded.email.has_value());
    assert(loaded.created_at == original.created_at);

    std::cout << "    PASS (serialized " << buffer.size() << " bytes)" << std::endl;
}

int main() {
    std::cout << "=== Test Case 09: SQLite Schema Types ===" << std::endl;

    test_user_creation();
    test_post_creation();
    test_comment_creation();
    test_optional_fields();
    test_post_status_enum();
    test_binary_serialization();
    test_binary_optional_null();

    std::cout << "=== All tests passed! ===" << std::endl;
    return 0;
}
